parser/protoc/parser.go

```diff
@@ -0,0 +1,183 @@
+		p.protoPaths = append(p.protoPaths, path...)
+	}
+}
+
+// AddDescriptorSetIn adds `--descriptor_set_in` to `protoc`.
+func AddDescriptorSetIn(files ...*descriptorpb.FileDescriptorProto) Option {
+	return func(p *Parser) {
+		p.descriptorSetIn = append(p.descriptorSetIn, files...)
+	}
+}
+
+// ExcludeCommonProtos excludes pre-loaded, common protobuf files.
+func ExcludeCommonProtos() Option {
+	return func(p *Parser) {
+		p.excludeCommonProtos = true
+	}
+}
+
+// Command sets the `protoc` command line path.
+func Command(path string) Option {
+	return func(p *Parser) {
+		p.command = path
+	}
+}
+
+// Parse parses a list of protobuf file with the given Options.
+func (p *Parser) Parse(files ...string) (*descriptorpb.FileDescriptorSet, error) {
+	return p.parse(files...)
+}
+
+func (p *Parser) parse(files ...string) (*descriptorpb.FileDescriptorSet, error) {
+	args := []string{
+		"--include_source_info",
+		"--proto_path=.",
+	}
+	if p.includeImports {
+		args = append(args, "--include_imports")
+	}
+
+	for _, path := range p.protoPaths {
+		args = append(args, "--proto_path="+path)
+	}
+
+	fileDescSet := &descriptorpb.FileDescriptorSet{}
+	fileDescSet.File = append(fileDescSet.File, p.descriptorSetIn...)
+	if !p.excludeCommonProtos {
+		imports, err := extractImportsFromFiles(files...)
+		if err != nil {
+			return nil, err
+		}
+		for _, name := range imports {
+			if f, err := protos.FindByName(name); err == nil {
+				fileDescSet.File = append(fileDescSet.File, f)
+			}
+		}
+	}
+	if len(fileDescSet.GetFile()) > 0 {
+		descSetInFile, err := ioutil.TempFile("", "protoc_compiler")
+		if err != nil {
+			return nil, err
+		}
+		defer os.Remove(descSetInFile.Name())
+		writeFileDescriptorSet(descSetInFile.Name(), fileDescSet)
+		args = append(args, "--descriptor_set_in="+descSetInFile.Name())
+	}
+
+	tempOutFile, err := ioutil.TempFile("", "protoc_compiler")
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(tempOutFile.Name())
+	args = append(args, "--descriptor_set_out="+tempOutFile.Name())
+
+	args = append(args, files...)
+
+	cmd := exec.Command(p.command, args...)
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return nil, fmt.Errorf("Running %s: %s, %v", strings.Join(cmd.Args, " "), string(out), err)
+	}
+
+	b, err := ioutil.ReadFile(tempOutFile.Name())
+	if err != nil {
+		return nil, err
+	}
+	out := &descriptorpb.FileDescriptorSet{}
+	if err := proto.Unmarshal(b, out); err != nil {
+		return nil, err
+	}
+	return out, err
+}
+
+func extractImportsFromFiles(files ...string) ([]string, error) {
+	s := make(map[string]struct{})
+	for _, f := range files {
+		b, err := ioutil.ReadFile(f)
+		if err != nil {
+			return nil, err
+		}
+		names := extractImports(string(b))
+		for _, name := range names {
+			s[name] = struct{}{}
+		}
+	}
+	var names []string
+	for key := range s {
+		names = append(names, key)
+	}
+	return names, nil
+}
+
+var importPattern = regexp.MustCompile(`\s*import\s+"([a-zA-Z0-9/\.]+)"\s*;`)
```

Comment:
Are `-` or `_` allowed in filenames? Also same issue as before with `(?:^|\s+)` probably being a better way to start.

In general, I'm not crazy about using regexes to pull these out. I think we should try not to do any manual parsing if possible.

Reply:
Actually, we always have this issue because of multiline comments
```
/**
OLD CODE FOR REFERENCE

package foo;

import "blah.proto";

message Bar {}

NEW CODE BELOW:
*/

package foo.v2;

import "blah/v2.proto";

message BarV2 {}
```

Reply:
Good points, but it is safe to add unnecessary dependencies in `protoc`. In this case, it is Okay to search and add `blah.proto` if found to the dependencies.